ipn/ipnlocal/local.go

```diff
@@ -3522,18 +3523,106 @@ func (b *LocalBackend) setTCPPortsInterceptedFromNetmapAndPrefsLocked() {
 				}
 			}
 			if b.serveConfig.Valid() {
+				servePorts := make([]uint16, 0, 3)
 				b.serveConfig.TCP().Range(func(port uint16, _ ipn.TCPPortHandlerView) bool {
 					if port > 0 {
-						handlePorts = append(handlePorts, uint16(port))
+						servePorts = append(servePorts, uint16(port))
 					}
 					return true
 				})
+				handlePorts = append(handlePorts, servePorts...)
+				b.updateServeTCPPortNetMapAddrListeners(servePorts)
 			}
 		}
 	}
 	b.setTCPPortsIntercepted(handlePorts)
 }
 
+// updateServeTCPPortNetMapAddrListeners starts a net.Listen for configured
+// Serve ports on all the node's addresses.
+// Existing Listeners are closed if port no longer in incoming ports list.
+//
+// b.mu must be held.
+func (b *LocalBackend) updateServeTCPPortNetMapAddrListeners(ports []uint16) {
+	nm := b.netMap
+	if nm == nil {
+		b.logf("netMap is nil")
+		return
+	}
+	if nm.SelfNode == nil {
+		b.logf("netMap SelfNode is nil")
+		return
+	}
+
+	if b.serveListeners == nil {
+		b.serveListeners = make(map[uint16]map[string]net.Listener)
+	}
+
+	for p, ls := range b.serveListeners {
+		if !slices.Contains(ports, p) {
+			for a, l := range ls {
+				b.logf("closing listener for port %d", p)
+				l.Close()
+				delete(ls, a)
+			}
+		}
+	}
+
+	for _, a := range nm.Addresses {
+		for _, p := range ports {
+			port := p
+			ipAddr := a.Addr()
+			network := "tcp"
+			if ipAddr.Is6() {
+				network = "tcp6"
+			}
+			address := net.JoinHostPort(ipAddr.String(), strconv.Itoa(int(p)))
+
+			if lm, ok := b.serveListeners[port]; ok {
+				if _, ok := lm[address]; ok {
+					continue // already listening
+				}
+			}
+
+			l, err := net.Listen(network, address)
+			if err != nil {
+				b.logf("error listening: %s, %s", network, address)
+				return
+			}
+
+			lm := b.serveListeners[port]
+			if lm == nil {
+				lm = make(map[string]net.Listener)
+				b.serveListeners[port] = lm
+			}
+			lm[address] = l
+
+			go func() {
```

Comment:
we prefer long-lived goroutines to have nice names in stacks, so move this to a method rather than an anonymous func (closure? is it closing over anything? hard to tell at quick glance with the inner and outer func being big enough)

Reply:
Thanks for the context!